src/bscPlugin/validation/ScopeValidator.ts

```diff
@@ -27,181 +27,133 @@ const platformComponentNames = new Set(Object.values(components).map(x => x.name
  */
 export class ScopeValidator {
 
-    private events: OnScopeValidateEvent[] = [];
+    /**
+     * The event currently being processed. This will change multiple times throughout the lifetime of this validator
+     */
+    private event: OnScopeValidateEvent;
 
     public processEvent(event: OnScopeValidateEvent) {
-        this.events.push(event);
-        event.scope.linkSymbolTable();
-        this.detectDuplicateEnums(event);
-        this.validateCreateObjectCalls(event);
-        this.iterateExpressions(event);
-        event.scope.unlinkSymbolTable();
+        this.event = event;
+        this.walkFiles();
+        this.detectDuplicateEnums();
     }
 
     public reset() {
+        this.event = undefined;
         this.onceCache.clear();
         this.multiScopeCache.clear();
-        this.events = [];
     }
 
-    /**
-     * Adds a diagnostic to the first scope for this key. Prevents duplicate diagnostics
-     * for diagnostics where scope isn't important. (i.e. CreateObject validations)
-     */
-    private addDiagnosticOnce(event: OnScopeValidateEvent, diagnostic: BsDiagnostic) {
-        this.onceCache.getOrAdd(`${diagnostic.code}-${diagnostic.message}-${util.rangeToString(diagnostic.range)}`, () => {
-            event.scope.addDiagnostics([diagnostic]);
-            return true;
+    private walkFiles() {
+        this.event.scope.enumerateOwnFiles((file) => {
+            if (isBrsFile(file)) {
+                this.iterateFileExpressions(file);
+                this.validateCreateObjectCalls(file);
+            }
         });
     }
-    private onceCache = new Cache<string, boolean>();
 
-    private addDiagnostic(event: OnScopeValidateEvent, diagnostic: BsDiagnostic) {
-        event.scope.addDiagnostics([diagnostic]);
-    }
+    private expressionsByFile = new Cache<BrsFile, Readonly<ExpressionInfo>[]>();
+    private iterateFileExpressions(file: BrsFile) {
+        const { scope } = this.event;
+        //build an expression collection ONCE per file
+        const expressionInfos = this.expressionsByFile.getOrAdd(file, () => {
```

Comment:
This is the core of the performance improvement. The expression info is computed _once_ for each file (previously this would happen once per scope per file).